crates/curp/src/server/gc.rs

```diff
@@ -23,25 +23,23 @@ pub(super) async fn gc_cmd_board<C: Command>(
             _ = shutdown_listener.wait() => break,
         }
         let mut board = cmd_board.write();
-
-        // last_check_len_xxx should always be smaller than board.xxx_.len(), the check is just for precaution
-
-        if last_check_len_er <= board.er_buffer.len() {
-            let new_er_buffer = board.er_buffer.split_off(last_check_len_er);
-            board.er_buffer = new_er_buffer;
-            last_check_len_er = board.er_buffer.len();
-        }
-
-        if last_check_len_asr <= board.asr_buffer.len() {
-            let new_asr_buffer = board.asr_buffer.split_off(last_check_len_asr);
-            board.asr_buffer = new_asr_buffer;
-            last_check_len_asr = board.asr_buffer.len();
+        let expired_er_ids: Vec<_> = board
+            .er_buffer
+            .keys()
+            .copied()
+            .filter(|ProposeId(client_id, _)| !lease_mamanger.read().check_alive(*client_id))
+            .collect();
+        for id in expired_er_ids {
+            let _ignore = board.er_buffer.swap_remove(&id);
         }
-
-        if last_check_len_conf <= board.conf_buffer.len() {
-            let new_conf = board.conf_buffer.split_off(last_check_len_conf);
-            board.conf_buffer = new_conf;
-            last_check_len_conf = board.conf_buffer.len();
+        let expired_asr_ids: Vec<_> = board
+            .asr_buffer
+            .keys()
+            .copied()
+            .filter(|ProposeId(client_id, _)| !lease_mamanger.read().check_alive(*client_id))
+            .collect();
+        for id in expired_asr_ids {
+            let _ignore = board.asr_buffer.swap_remove(&id);
         }
     }
```

Comment:
how about some code like:
```rust
let remove_expired = |buf: &mut IndexMap<
    ProposeId,
    Result<<C as Command>::ER, <C as Command>::Error>,
>| {
    buf.keys()
        .filter(|ProposeId(client_id, _)| !lease_mamanger.read().check_alive(*client_id))
        .for_each(|id| {
            _ = board.asr_buffer.swap_remove(&id);
        });
};
remove_expired(&mut board.er_buffer);
remove_expired(&mut board.asr_buffer);
```
I think we don't need a `copied` here because we only uses `&ProposeId`.

Reply:
THe `swap_remove` requires mutable borrow of the buf. We needs to collect the ids before any remove.

Reply:
ref: https://docs.rs/indexmap/latest/indexmap/map/struct.IndexMap.html#method.retain